pkg/oomstore/apply.go

```diff
@@ -0,0 +1,192 @@
+package oomstore
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/oom-ai/oomstore/internal/database/metadata"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types/apply"
+	"gopkg.in/yaml.v3"
+)
+
+func (s *OomStore) Apply(ctx context.Context, opt apply.ApplyOpt) error {
+	data := make(map[string]interface{})
+	if err := yaml.NewDecoder(opt.R).Decode(data); err != nil {
+		return err
+	}
+
+	var kind string
+	if k, ok := data["kind"]; ok {
+		kind = k.(string)
+	} else {
+		return fmt.Errorf("Invalid data missing variable kind")
+	}
+
+	switch strings.ToLower(kind) {
+	case "entity":
+		var entity apply.Entity
+		if err := mapstructure.Decode(data, &entity); err != nil {
+			return err
+		}
+		return s.metadata.WithTransaction(ctx, func(c context.Context, tx metadata.WriteStore) error {
+			return s.applyEntity(ctx, tx, entity)
+		})
+	case "feature":
+		var feature apply.Feature
+		if err := mapstructure.Decode(data, &feature); err != nil {
+			return err
+		}
+
+		group, err := s.metadata.GetGroupByName(ctx, feature.GroupName)
+		if err != nil {
+			return err
+		}
+		feature.GroupID = group.ID
+
+		return s.metadata.WithTransaction(ctx, func(c context.Context, tx metadata.WriteStore) error {
+			return s.applyFeature(ctx, tx, feature)
+		})
+	case "featuregroup", "group":
+		var group apply.FeatureGroup
+		if err := mapstructure.Decode(data, &group); err != nil {
+			return err
+		}
+		entity, err := s.metadata.GetEntityByName(ctx, group.EntityName)
+		if err != nil {
+			return err
+		}
+		group.EntityID = entity.ID
+
+		return s.metadata.WithTransaction(ctx, func(c context.Context, tx metadata.WriteStore) error {
+			return s.applyGroup(ctx, tx, group)
+		})
+	default:
+		return fmt.Errorf("invalid kind %s", kind)
+	}
+}
+
+func (s *OomStore) applyEntity(ctx context.Context, txStore metadata.WriteStore, newEntity apply.Entity) error {
```

Comment:
txStore -> store? Since the method allows non tx store to be passed in

Reply:
Not sure I understand...  but here the methods that require transactions are called through the txStore  and the non-tx methods are not called using the txStore

Reply:
🆗